### ISSUE TITLE
icc error: non-pointer conversion from "int" to "unsigned short" may lose significant bits

### ISSUE THREAD

**Issue by mathomp4** [curl/curl#9156]
### I did this

```
  ./configure --enable-warnings --enable-werror CC=icc --enable-debug --without-ssl
```

using Intel oneAPI icc as the compiler in [my attempt to add Intel into curl CI](https://github.com/mathomp4/curl/pull/1).

The latest attempt failed with:

```
libtool: compile:  icc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -DCURL_HIDDEN_SYMBOLS -we140,147,165,266 -wd279,981,1469 -Wall -w2 -Wcheck -Wcomment -Wdeprecated -Wmissing-prototypes -Wp64 -Wpointer-arith -Wreturn-type -Wshadow -Wuninitialized -Wunused-function -fvisibility=hidden -std=gnu89 -g -O0 -fno-omit-frame-pointer -fno-strict-aliasing -fp-model precise -shared-intel -pthread -Werror -MT libcurl_la-curl_sasl.lo -MD -MP -MF .deps/libcurl_la-curl_sasl.Tpo -c curl_sasl.c  -fPIC -DPIC -o .libs/libcurl_la-curl_sasl.o
icc: command line remark #10010: option '-we140,147,165,266' is deprecated and will be removed in a future release. See '-help deprecated'
icc: command line remark #10010: option '-wd279,981,1469' is deprecated and will be removed in a future release. See '-help deprecated'
curl_sasl.c(357): error #2259: non-pointer conversion from "int" to "unsigned short" may lose significant bits
    enabledmechs = sasl->authmechs & sasl->prefmech;
                 ^
compilation aborted for curl_sasl.c (code 2)
make[2]: *** [Makefile:2179: libcurl_la-curl_sasl.lo] Error 1
```

I tried to see if I could duplicate this on the cluster where I have Intel, but it apparently doesn't have the right things to build this file (gsasl?)

### I expected the following

For curl to build

### curl/libcurl version

This should be equivalent to [4c46c82](https://github.com/mathomp4/curl/commit/4c46c829f5ed7c37b3313c67a6662b7332c1ff7f)

as I merged in the latest `master` before re-submitting CI

### operating system

An ubuntu-latest GitHub Actions VM

**Comment by bagder** [comment]
isn't that an expression that involves three variables that are all `unsigned short` ?

**Comment by mathomp4** [comment]
> isn't that an expression that involves three variables that are all `unsigned short` ?

Huh. You are right. That is...weird. From looking online, it almost looks like this is a bug/feature of the Intel C Compiler (see https://community.intel.com/t5/Intel-C-Compiler/Remark-2259/td-p/888954 and https://community.intel.com/t5/Intel-C-Compiler/compilation-error-with-ntohs-htons-function-with-icc-under-linux/m-p/797937)

I wonder if there's a way we could just keep that error a warning?

**Comment by MarcelRaad** [comment]
As the result of `(sasl->authmechs & sasl->prefmech)` is promoted to an `int`, icc is technically correct, even if the warning doesn't make much sense here. Explicitly casting the result to `unsigned short` should fix it, shouldn't it?

**Comment by mathomp4** [comment]
@MarcelRaad 

Sooo...let's say you were a primarily a Fortran programmer and could write Hello World in C but not much else. Would this be:

```c
    enabledmechs = (unsigned short)(sasl->authmechs & sasl->prefmech);
```

?

**Comment by MarcelRaad** [comment]
@mathomp4 Yes 👍

**Comment by mathomp4** [comment]
> @mathomp4 Yes 👍

Okay. I'll try that in my PR on my fork...

**Comment by bagder** [comment]
> As the result of `(sasl->authmechs & sasl->prefmech)` is promoted to an `int`, icc is technically correct, even if the warning doesn't make much sense here. Explicitly casting the result to `unsigned short` should fix it, shouldn't it?

It will however leave a bad taste in my mouth since it adds a typecast that shouldn't even have to be there. If we cannot figure out a way around it, I think it warrants an added comment that explains why it is there.

**Comment by bagder** [comment]
> enumerated type mixed with another type

Those warnings/errors should be fixed in master now.

**Comment by mathomp4** [comment]
> > maybe I should just make the PR here and then you/we can work on getting Intel happy via that?
> 
> Sounds like a plan!

Done:

https://github.com/curl/curl/pull/9179

**Comment by bagder** [comment]
Fixed in #9179

**Comment by mathomp4** [comment]
@bagder Can confirm that current `master` builds with my library stack using Intel compilers! Thanks for all the work!
